File: library/src/main/java/com/accesium/sendtopush/service/GcmRegistrationService.java
package com.accesium.sendtopush.service;

import com.accesium.sendtopush.tools.Log;
import com.google.android.gms.gcm.GoogleCloudMessaging;

import java.io.IOException;

import rx.Observable;
import rx.Subscriber;
import rx.schedulers.Schedulers;

/**
 * Created by Fran Gilberte on 19/01/2016.
 */
public class GcmRegistrationService {
    private static GoogleCloudMessaging gcm;

    public GcmRegistrationService(GoogleCloudMessaging gcm) {
        this.gcm = gcm;
    }

    public Observable<String> register(String gcmSenderId) {
        return Observable.create((Subscriber<? super String> subscriber) -> {
            try {
                if (!subscriber.isUnsubscribed()) {
                    String token = gcm.register(gcmSenderId);

                    subscriber.onNext(token);
                    Log.d("Gcm register success");

                    subscriber.onCompleted();
                }
            } catch (IOException e) {
                subscriber.onError(e);
                Log.d("Gcm register failed");
            }
        }).subscribeOn(Schedulers.io());
    }

    public Observable<Boolean> unregister() {
        return Observable.create((Subscriber<? super Boolean> subscriber) -> {
            try {
                if (!subscriber.isUnsubscribed()) {
                    gcm.unregister();
                    subscriber.onNext(true);
                    Log.d("Gcm unregister success");
                    subscriber.onCompleted();
                }
            } catch (IOException e) {
                subscriber.onError(e);
                Log.d("Gcm unregister failed");
            }
        }).subscribeOn(Schedulers.io());
    }

}

File: library/src/main/java/com/accesium/sendtopush/service/GcmRegistrationService.kt
package com.accesium.sendtopush.service

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.iid.FirebaseInstanceIdService
import io.reactivex.subjects.BehaviorSubject

/**
 * Created by Fran Gilberte on 19/01/2016.
 */
class GcmRegistrationService : FirebaseInstanceIdService() {

    companion object {
        var subject = BehaviorSubject.create<String>()
        var token = ""
        var tokenReceived = false

        fun register(context: Context): BehaviorSubject<String> {
            FirebaseApp.initializeApp(context)
            token = FirebaseInstanceId.getInstance().token ?: ""
            if(token.isNotBlank()) {
                subject.onNext(token)
                tokenReceived = true
            } else {
                tokenReceived = false
            }
            return subject
        }

        fun unregister() {
            Thread{ FirebaseInstanceId.getInstance().deleteInstanceId() }.start()
        }
    }

    override fun onTokenRefresh() {
        // Get updated InstanceID token.
        FirebaseInstanceId.getInstance().token?.let {
            token = it
            if(!tokenReceived) {
                subject.onNext(it)
                tokenReceived = true
            }
        }
    }


}
